tools/get_training_info/get_training_info.py

```diff
@@ -0,0 +1,456 @@
+        args.tensor_model_parallel_size
+    while (vocab_size % multiple) != 0:
+        vocab_size += 1
+    return vocab_size
+
+
+def get_ffn_hidden_size(args):
+    return int((4 * args.hidden_size * 2 / 3) / 64) * 64
+
+
+def get_megatron_info(disable_profile=False):
+    args = get_args()
+    h = args.hidden_size
+    t = args.tensor_model_parallel_size
+    n = args.num_layers
+    p = args.pipeline_model_parallel_size
+    d = args.world_size // p // t
+    v = get_padded_vocab_size(args)
+    s = args.seq_length
+    b = args.micro_batch_size
+    a = args.num_attention_heads
+    ga = args.global_batch_size // args.micro_batch_size // d
+    f = get_ffn_hidden_size(args)
+    per_stage_layer_num = n // p
+
+    # parameters
+    total_parameters = v * h + s * h + (7 * h + 4 * h * h + 3 * f * h +
+                                        2 * f) * n
+    total_parameters_formatted = f'{total_parameters:,}'
+
+    total_parameters_per_gpu = v * h / t + s * h + (
+        7 * h + 4 * h * h / t + 3 * f * h / t + 2 * f) * per_stage_layer_num
+    total_parameters_per_gpu_formatted = f'{int(total_parameters_per_gpu):,}'
+
+    activations = n * (10 * s * h * b +
+                       4 * s * b * f) + 1.5 * s * b * h + 4.5 * s * b * v
+    activations_per_gpu = (per_stage_layer_num * (
+        (5 * s * h * b + 4 * s * b * f) / t + 5 * s * h * b) +
+                           1.5 * s * b * h) * p
+
+    # communication
+    pp_comm_count = 0 if p == 1 else args.world_size * ga * 4
+    pp_comm_size = pp_comm_count * s * b * h
+
+    tp_comm_count = 0 if t == 1 else args.world_size * 4 * per_stage_layer_num * ga
+    tp_comm_size = tp_comm_count * s * b * h
+
+    dp_comm_count = 0 if d == 1 else 2
+    dp_comm_size = total_parameters_per_gpu * 4 if args.bf16 else total_parameters_per_gpu * 2
+
+    total_comm_count = pp_comm_count + tp_comm_count + dp_comm_count
+    total_comm_size = pp_comm_size + tp_comm_size + dp_comm_size
+
+    # peak memory
+    if args.bf16:
+        loss_logits_mem = 5 * s * b * v / t if p == 1 else 0
+        peak_mem = max(
+            memory_giga_bytes(total_parameters_per_gpu * (1 + 2 + 2 / d + 2)),
+            memory_giga_bytes(total_parameters_per_gpu * (1 + 2 + 6 / d) +
+                              activations_per_gpu + loss_logits_mem))
+        gradient = total_parameters * 2
+        gradient_per_gpu = total_parameters_per_gpu * 2
+        optimizer = total_parameters * 6
+        optimizer_per_gpu = total_parameters_per_gpu * 6 / d
+    elif args.fp16:
+        loss_logits_mem = 2 * s * b * v / t if p == 1 else 0
+        peak_mem = max(
+            memory_giga_bytes(total_parameters_per_gpu * (1 + 1 + 2 / d + 2)),
+            memory_giga_bytes(total_parameters_per_gpu * (1 + 1 + 8 / d) +
+                              activations_per_gpu + loss_logits_mem))
+        gradient = total_parameters
+        gradient_per_gpu = total_parameters_per_gpu
+        optimizer = total_parameters * 8
```

Comment:
why `optimizer = total_parameters * 6`? For each parameter in the model, AdamW keeps track of First Moment Vector and Second Moment Vector.As a result, the CUDA memory requirement for using the AdamW optimizer is approximately 2 times the memory required for the model parameters themselves.

Reply:
Thanks for the super fast reply! I meant line 228, about the factor `(1 + 2 + 2 / d + 2)`. The first terms `1` and `2` are params in bfloat16 and gradients in fp32 but I cannot derive `2/d +2`. Are those the optimizer states distributed in some way? Where does the second term `2` come from?